File: 1_Linear_Regression/linear_regression.R


###############################################
## Function 1: Plain version of Gauss Jordan ##
###############################################


myGaussJordan <- function(A, m){
  
  # Perform Gauss Jordan elimination on A.
  # 
  # A: a square matrix.
  # m: the pivot element is A[m, m].
  # Returns a matrix with the identity matrix 
  # on the left and the inverse of A on the right. 

  #############################################
  ## FILL IN THE BODY OF THIS FUNCTION BELOW ##
  #############################################
  n <- dim(A)[1]
  B <- cbind(A, diag(rep(1,n)))
  for (k in 1:m) {                               #m is row number
    a <- B[k,k]
    for (j in 1:(n * 2)) {                 
      B[k,j] <- B[k, j] / a                      #divide the whole row by pivot
    }
    for (i in 1:n)                               #n is col, for each col
      if (i != k) {
        b <- B[i,k]
        for (j in 1 : (n * 2))
        B[i,j] <-B[i, j] - b* B[k,j]
      }
  
  ## Function returns the matrix B
  }
  return(B)
  
}

####################################################
## Function 2: Vectorized version of Gauss Jordan ##
####################################################

myGaussJordanVec <- function(A, m){
  
  # Perform Gauss Jordan elimination on A.
  # 
  # A: a square matrix.
  # m: the pivot element is A[m, m].
  # Returns a matrix with the identity matrix 
  # on the left and the inverse of A on the right.
  
  #############################################
  ## FILL IN THE BODY OF THIS FUNCTION BELOW ##
  #############################################
  n <- dim(A)[1]
  B <- cbind(A, diag(rep(1,n)))
  for(k in 1:m) {
    B[k,] <- B[k,] / B[k,k]
    for (i in 1:n){ 
      if (i != k) {
        B[i,] <- B[i,] - B[k,] * B[i,k]
      }
    }
  }
  ## Function returns the matrix B
  return(B)
  
}



######################################################
## Function 3: Linear regression using Gauss Jordan ##
######################################################

myLinearRegression <- function(X, Y){
  
  # Find the regression coefficient estimates beta_hat
  # corresponding to the model Y = X * beta + epsilon
  # Your code must use one of the 2 Gauss Jordan 
  # functions you wrote above (either one is fine).
  # Note: we do not know what beta is. We are only 
  # given a matrix X and a vector Y and we must come 
  # up with an estimate beta_hat.
  # 
  # X: an 'n row' by 'p column' matrix of input variables.
  # Y: an n-dimensional vector of responses

  #############################################
  ## FILL IN THE BODY OF THIS FUNCTION BELOW ##
  #############################################
  
  ## Let me start things off for you...
  n <- nrow(X)
  p <- ncol(X)
  
  Z = cbind(rep(1,n), X, Y)
  A = t(Z)%*%Z
  S = myGaussJordan(A, p+1)                    # you find S, and  the bottom right before I is negative beta_hat
  
  last_row = c(S[p+2,])                        #take last row
  index = length(last_row) - 1                 # find the last number before I
  beta_hat = -c(last_row[(p+3):index])         #take the subvector to get beta_hat
  
  
  ## Function returns the (p+1)-dimensional vector 
  ## beta_hat of regression coefficient estimates
  return(beta_hat)
  
}

########################################################
## Optional examples (comment out before submitting!) ##
########################################################

testing_Linear_Regression <- function(){
  
  ## This function is not graded; you can use it to 
  ## test out the 'myLinearRegression' function 

  ## Define parameters
  n    <- 5
  p    <- 5
  
  ## Simulate data from our assumed model.
  ## We can assume that the true intercept is 0
  X    <- matrix(rnorm(n * p), nrow = n)
  beta <- matrix(1:p, nrow = p)
  Y    <- X %*% beta + rnorm(n)
  
  ## Save R's linear regression coefficients
  R_coef  <- coef(lm(Y ~ X))
  
  ## Save our linear regression coefficients
  my_coef <- myLinearRegression(X, Y)
  
  ## Are these two vectors different?
  sum_square_diff <- sum((R_coef - my_coef)^2)
  if(sum_square_diff <= 0.001){
    return('Both results are identical')
  }else{
    return('There seems to be a problem...')
  }
  
}

File: 1_Linear_Regression/linear_regression.py
import numpy as np

#Function 1: Plain version of Gauss Jordan
def myGaussJordan(A, m):
    """
    Perform Gauss Jordan elimination on A.
    
    A: a square matrix.
    m: the pivot element is A[m, m].
    Returns a matrix with the identity matrix 
    on the left and the inverse of A on the right. 
    """
    n = A.shape[0]
    B = np.hstack((A, np.identity(n)))
    for k in range(n):  # 修改为n，以确保处理所有行
        a = B[k, k]
        for j in range(n * 2):
            B[k, j] = B[k, j] / a
        for i in range(n):
            if i != k:
                b = B[i, k]
                for j in range(n * 2):
                    B[i, j] = B[i, j] - B[k, j] * b

    return B

def myGaussJordanVec(A, m):
    """
    Perform Gauss Jordan elimination on A using a vectorized approach.
    
    A: a square matrix.
    m: the pivot element is A[m, m].
    Returns a matrix with the identity matrix 
    on the left and the inverse of A on the right.
    """
    n = A.shape[0]
    B = np.hstack((A, np.identity(n)))
    for k in range(n):  # 修改为 n，确保处理所有行
        B[k, :] = B[k, :] / B[k, k]  # 缩放行使主对角线元素为 1
        for i in range(n):
            if i != k:
                B[i, :] = B[i, :] - B[k, :] * B[i, k]  # 消除第i行的第k列

    return B

def myLinearRegression(X, Y):
    """
    Find the regression coefficient estimates beta_hat
    corresponding to the model Y = X * beta + epsilon
    using Gauss Jordan elimination method.
    
    X: an 'n row' by 'p column' matrix (np.array) of input variables.
    Y: an n-dimensional vector (np.array) of responses
    """
    n, p = X.shape
    
    # Adding intercept term by appending a column of ones
    X = np.hstack([np.ones((n, 1)), X])
    
    # Building the normal equation components
    XTX = np.dot(X.T, X)
    XTY = np.dot(X.T, Y)
    
    # Augmenting XTX with XTY for Gauss-Jordan
    A = np.hstack([XTX, XTY.reshape(-1, 1)])
    
    # Applying Gauss-Jordan elimination
    S = myGaussJordanVec(A, p + 1)  # Using the vectorized Gauss Jordan function
    
    # The last column of S (after Gauss-Jordan) will be the solution beta_hat
    beta_hat = S[:, -1]  # Extracting beta_hat
    
    return beta_hat

# Note: Ensure to adjust the Gauss Jordan function to handle non-square matrices if needed.

  


########################################################
## Optional examples (comment out before submitting!) ##
########################################################

#def testing_Linear_Regression():
  
  ## This function is not graded; you can use it to 
  ## test out the 'myLinearRegression' function 
  #n = 100
  #p = 3
  #X = np.random.standard_normal((n*p)).reshape((n,p))
  
  #beta = np.arange(1,p+1,1).reshape((p,1))
    
  #Y = np.dot(X,beta) + np.random.standard_normal(n).reshape(n,1)

  ## You can set up a similar test function as was 
  ## provided to you in the R file.



